File: services/ai-gateway/src/polis_aigateway/audit.py
"""Best-effort audit emit — mirrors proof-service ``emitAudit`` (§26.3).

Uses ``urllib.request`` (stdlib) so the gateway has no HTTP client dependency.
Never raises: failures are logged to stderr.
"""

from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request
from typing import Any

_ACTOR: dict[str, str] = {"type": "service", "id": "ai-gateway"}


def emit_audit(
    *,
    event_type: str,
    action: str,
    target: dict[str, str],
    data: dict[str, Any],
    correlation_id: str | None = None,
) -> None:
    """POST an audit event to ``AUDIT_INTERNAL_URL/internal/audit/events``.

    Target keys are chosen so the existing BFF
    ``GET /api/v1/audit/:objectType/:objectId`` surfaces them:
      - ``ai.answer.requested`` → ``target:{type:'ai-trace', id:traceId}``
    """
    base = os.environ.get("AUDIT_INTERNAL_URL", "http://localhost:8600")
    body = {
        "eventType": event_type,
        "action": action,
        "visibility": "public",
        "actor": _ACTOR,
        "target": target,
        "data": data,
        "correlationId": correlation_id,
    }
    try:
        req = urllib.request.Request(
            base + "/internal/audit/events",
            data=json.dumps(body).encode(),
            headers={"content-type": "application/json"},
            method="POST",
        )
        urllib.request.urlopen(req, timeout=5)  # noqa: S310 — internal URL
    except (urllib.error.URLError, OSError, TimeoutError) as exc:
        print(f"[audit] emit failed: {exc}", file=sys.stderr)

File: services/contribution-service/src/audit.ts
import { fetchWithTimeout, internalHeaders } from '@polis/service-runtime';
import type { PublishDenial, RequestedCommitmentScope } from './representative-publish-gate.js';
import type { AuditEvent } from './types.js';

const AUDIT_TIMEOUT_MS = 5_000;

async function postAudit(event: AuditEvent): Promise<void> {
  const base = process.env.AUDIT_INTERNAL_URL ?? 'http://localhost:8600';
  const headers = internalHeaders();
  const response = await fetchWithTimeout(
    base + '/internal/audit/events',
    {
      method: 'POST',
      headers,
      body: JSON.stringify({
        eventType: event.eventType,
        action: event.action,
        visibility: event.visibility ?? 'public',
        actor: event.actor ?? { type: 'service', id: 'contribution-service' },
        target: event.target,
        data: event.data,
        correlationId: event.correlationId ?? null,
      }),
    },
    AUDIT_TIMEOUT_MS,
  );
  if (!response.ok) throw new Error(`audit-service returned ${response.status}`);
}

/** Existing post-mutation domain events remain best-effort, but failures are explicit. */
export async function emitAudit(event: AuditEvent): Promise<void> {
  try {
    await postAudit(event);
  } catch (error) {
    console.error(
      JSON.stringify({
        service: 'contribution-service',
        stage: 'audit-emit',
        warning: error instanceof Error ? error.message : 'unknown',
      }),
    );
  }
}

export class AuditUnavailableError extends Error {
  constructor() {
    super('required audit unavailable');
    this.name = 'AuditUnavailableError';
  }
}

export async function requiredAudit(event: AuditEvent): Promise<void> {
  try {
    await postAudit({ ...event, visibility: 'restricted' });
  } catch (error) {
    console.error(
      JSON.stringify({
        service: 'contribution-service',
        stage: 'audit-required',
        warning: error instanceof Error ? error.message : 'unknown',
      }),
    );
    throw new AuditUnavailableError();
  }
}

/**
 * Best-effort reward-eligibility emit. On approval of a non-political
 * contribution, POST the submission to rewards-service which evaluates the
 * §20.4 eligibility policy (ADR-007 + monthly cap). Failures (rewards-service
 * unreachable) are logged and never fail the originating approval — matches
 * emitAudit. Eligibility is reconciled by re-POSTing the submission to the
 * idempotent endpoint (dedupes on submission_id).
 */
export async function emitRewardEligibility(input: {
  submissionId: string;
  contributorId: string;
  contributionClass: string;
}): Promise<void> {
  const base = process.env.REWARDS_INTERNAL_URL ?? 'http://localhost:8460';
  try {
    await fetch(base + '/internal/rewards/eligibility', {
      method: 'POST',
      headers: internalHeaders(),
      body: JSON.stringify({
        submissionId: input.submissionId,
        contributorId: input.contributorId,
        contributionClass: input.contributionClass,
      }),
    });
  } catch (error) {
    console.error(
      JSON.stringify({
        service: 'contribution-service',
        stage: 'reward-eligibility-emit',
        warning: error instanceof Error ? error.message : 'unknown',
      }),
    );
  }
}

export async function emitRepresentativeDeniedAudit(input: {
  eventType: string;
  action: string;
  target: { type: string; id: string };
  citizenId: string;
  mandateHolderId: string;
  denial: PublishDenial;
  requestedScope: RequestedCommitmentScope;
}): Promise<void> {
  await emitAudit({
    eventType: input.eventType,
    action: input.action,
    target: input.target,
    visibility: 'restricted',
    data: {
      citizenId: input.citizenId,
      mandateHolderId: input.mandateHolderId,
      denied: true,
      reason: input.denial.reason,
      error: input.denial.error,
      field: input.denial.field ?? null,
      requestedScope: input.requestedScope,
    },
  });
}
